### PR TITLE
WIXBUG4931 Fix drawing of image static controls

### PR DESCRIPTION
**Pull request by drolevar**
Fixes issue https://github.com/wixtoolset/issues/issues/4931

Static image control has no reason to respond to WM_UPDATEUISTATE.
If it does, it redraws itself, but not the controls which might be on top of it.
I've created a superclass for the image control which ignores WM_UPDATEUISTATE message.

The idea was taken from here:
https://devblogs.microsoft.com/oldnewthing/20161212-00/?p=94915

### CODE REVIEW

File: src/dutil/thmutil.cpp [hunk 1]
```diff
@@ -40,13 +40,16 @@ const DWORD GROW_FONT_INSTANCES = 3;
...
+const LPCWSTR THEME_WC_IMAGE = L"ThemeImage";
 
 static Gdiplus::GdiplusStartupInput vgsi;
 static Gdiplus::GdiplusStartupOutput vgso = { };
 static ULONG_PTR vgdiToken = 0;
 static ULONG_PTR vgdiHookToken = 0;
 static HMODULE vhHyperlinkRegisteredModule = NULL;
 static HMODULE vhPanelRegisteredModule = NULL;
+static HMODULE vhImageRegisteredModule = NULL;
+static WNDPROC vpfnImageWndProc = NULL;
```
Comment (rseanhall) [inline src/dutil/thmutil.cpp:52]: This should probably be `vpfnImageBaseWndProc` or `vpfnImageParentWndProc`.

File: src/dutil/thmutil.cpp [hunk 2]
```diff
@@ -1598,6 +1614,8 @@ static HRESULT RegisterWindowClasses(
     HRESULT hr = S_OK;
     WNDCLASSW wcHyperlink = { };
     WNDCLASSW wcPanel = { };
+    WNDCLASSW wcImage = { };
+    WNDPROC pfnImageWndProc;
```
Comment (rseanhall) [inline src/dutil/thmutil.cpp:1618]: The WiX style is to initialize all variables where they are declared, so set this to `NULL` here.

File: src/dutil/thmutil.cpp [hunk 3]
```diff
@@ -4919,6 +4953,17 @@ static LRESULT CALLBACK PanelWndProc(
...
+static LRESULT CALLBACK ImageWndProc(
+    __in HWND hWnd,
+    __in UINT uMsg,
+    __in WPARAM wParam,
+    __in LPARAM lParam
+    )
+{
+    return uMsg == WM_UPDATEUISTATE
+        ? ::DefWindowProc(hWnd, uMsg, wParam, lParam)
+        : (*vpfnImageWndProc)(hWnd, uMsg, wParam, lParam);
```
Comment (rseanhall) [inline src/dutil/thmutil.cpp:4965]: I think we should use a `switch` statement here so it's easier to modify in the future.

**Review by rseanhall — COMMENTED** [review]
Thanks! The progress bar also needs this new class when it is a static control.

**Comment by drolevar** [comment]
> Thanks! The progress bar also needs this new class when it is a static control.

@rseanhall What's the specific issue with the progress bar? Should I also fix it as a part of this PR?

**Comment by rseanhall** [comment]
Yes. This issue is for all static controls that are owner drawn. The progress bar can be a static control that is owner drawn. https://github.com/wixtoolset/dutil/blob/d224c628ef4964c5ba29dc66a83970460205cf21/src/dutil/thmutil.cpp#L5084-L5089

**Comment by drolevar** [comment]
@rseanhall So the solution would be to use that same window class also for progress bar?
But then I guess the class has to be called something else, like "ThemeOwnerDraw" for example?
```
        case THEME_CONTROL_TYPE_PROGRESSBAR:
            if (pControl->hImage || (pTheme->hImage && 0 <= pControl->nSourceX && 0 <= pControl->nSourceY))
            {
                wzWindowClass = THEME_WC_OWNERDRAW; // no such thing as an owner drawn progress bar so we'll make our own out of a static control.
                dwWindowBits |= SS_OWNERDRAW;
                pControl->dwInternalStyle |= INTERNAL_CONTROL_STYLE_OWNER_DRAW;
            }
```

**Comment by rseanhall** [comment]
I thought about that but it's owner draw for a static control so `THEME_WC_STATICOWNERDRAW` (or `THEME_WC_OWNERDRAWSTATIC`) which is kind of long. I figured it wasn't really going to be for anything except for images so I originally didn't say anything.

**Comment by drolevar** [comment]
@rseanhall Btw, two more questions:
- Will it be possible to accept it into wix3 since it's really annoying for a long of people and wix4 release date is not known yet?
- Is there a proper building/contributor guide for wix4?

**Comment by rseanhall** [comment]
1. That would be something to discuss in our online meeting that happens every other week. You should add a comment to the issue with why you think we need to take in v3, and then I could assign the issue to you and label it for discussion in the next meeting (likely 2021-02-18 at 9:30am Pacific).
2. There's the readme in the [Home repo](https://github.com/wixtoolset/Home) and this thread on wix-devs http://lists.wixtoolset.org/pipermail/wix-devs-wixtoolset.org/2021-January/001923.html.

**Comment by drolevar** [comment]
@rseanhall Is it OK as it is for this branch? If it is, I'll merge it to wix3, create a PR and make a comment for the issue as you suggested.

**Comment by rseanhall** [comment]
`THEME_WC_OWNERDRAW` is too generic. For example, we are doing owner draw for buttons already in some cases. I'm fine with `THEME_WC_IMAGE`, `THEME_WC_STATICOWNERDRAW`, or `THEME_WC_OWNERDRAWSTATIC`.

**Pull request merged**